mountpoint-s3/src/prefetch.rs

```diff
@@ -483,19 +473,33 @@ where
                     self.next_sequential_read_offset = next_request.end_offset();
                 }
             }
-            // We checked there was an inflight task that contained the target offset, so this
-            // is impossible.
-            assert!(self.current_task.is_some());
+            if self.current_task.is_none() {
+                // No inflight task containing the target offset.
+                trace!(current_offset=?self.next_sequential_read_offset, requested_offset=?offset, "seek failed: not enough inflight data");
+                return Ok(false);
+            }
             // We could try harder to preserve the backwards seek buffer if we're near the
             // request boundary, but it's probably not worth the trouble.
             self.backward_seek_window.clear();
         }
-        let mut seek_distance = offset - self.next_sequential_read_offset;
 
         let current_task = self
             .current_task
             .as_mut()
             .expect("a request existed that covered this seek offset");
+        // If we have enough bytes already downloaded (`available`) to skip straight to this read, then do
+        // it. Otherwise, we're willing to wait for the bytes to download only if they're coming "soon", where
+        // soon is defined as up to `max_forward_seek_distance` bytes ahead of the available offset.
+        let available_offset = current_task.available_offset();
+        if offset > available_offset.saturating_add(self.config.max_forward_seek_distance) {
```

Comment:
`max_forward_seek_distance` is now the maximum number of bytes we are happy to wait to be downloaded when seeking forward. Can we think of a more appropriate name?

Reply:
`max_forward_seek_wait_distance`?

Also, there's a subtle thing here that it doesn't matter if `available_offset + max_forward_seek_wait_distance > current_task.end_offset()`, because the loop above guarantees that the `offset` lands within this task in the loop above. Maybe worth a comment on the line above this one that selects `current_task`?

Reply:
oh, yeah, not regarding this particular line, but I also thought about the case when `offset` is beyond the last task's range, I'll add a comment and will try to add a test (if possible without too much hacks) 

`max_forward_seek_wait_distance` looks good to me, can not think about anything less wordy